Fix aliasing, metadata and edge case handling for FunctionalTensor, introduce some new wrapper tensor APIs

bdhirsh opened the pull request:
This PR provides some updates to the "wrapper tensor" subclass API, that made it possible to integrate with AOTAutograd (next PR in the stack).

Enumeration of changes:

**Changes that are general to wrapper subclasses**

(1) When a subclass specifies `dispatch_sizes_strides_policy="sizes"`, previously it would always cache the result of the first call somewhere on the python tensor. I removed this caching - if we cache, then we won't notice when there is a metadata mutation to the tensor.

(2) Allow subclasses using `dispatch_sizes_strides_policy="sizes"` to work with dynamic shapes. Previously, you'd get an error if you specified a wrapper subclass like the above and tried to give it dynamic sizes. From the comments this sounded like a coverage gap, so I removed the restriction. It also required removing an `as_int_unchecked` call from `ConcretePyInterpreterVTable::sym_sizes`.

(3) Added a new API for wrapper subclasses: `torch.utils._python_dispatch.return_and_correct_aliasing(func, args, out)`.

This API does the work of both (a) aliasing any output storages to inputs when it sees a view op, and (b) making sure that we return inputs directly, when we encounter an inplace op. This should be a one-liner that wrapper subclasses can add to their __torch_dispatch__ code in order to get correct aliasing.

Importantly, a wrapper subclass will **need** to use this API in their subclass to work properly with `torch.compile`.

**Changes specific to FunctionalTensor**

Trying to integrate with AOTAutograd exposed a few bugs in my subclass as written. Stuff that I had to fix included:

(1) Make `torch._sync()` work with `FunctionalTensor`, and make sure to temporarily disable `FunctionalTensorMode` while performing the sync

(2) Use `dispatch_sizes_strides_policy="sizes"`, and have every call to metadata get plumbed to the inner tensor. This is exactly what the C++ `FunctionalTensorWrapper` does today too. Although @ezyang points out that at some point we may want to "not trust" the strides of the inner tensor, and compute strides ourselves (FunctionalTensorWrapper... sort of did this a while ago, but we killed it everywhere except for XLA).

(3) Use the new `return_and_correct_aliasing` API.

(4) Make sure to include `DispatchKey.Functionalize` in the TLS, when redispatching in our mode. This ensures that factory ops are hit.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105236
* #104974
* #104483
* #106406
* __->__ #106405
* #106404
* #104482
* #107415
* #107416
* #107417

bdhirsh committed:
Fix aliasing, metadata and edge case handling for FunctionalTensor, introduce some new wrapper tensor APIs

[ghstack-poisoned]

soulitzer commented on torch/csrc/PyInterpreter.cpp:
```diff
@@ -663,8 +663,7 @@ c10::SymIntArrayRef ConcretePyInterpreterVTable::sym_sizes(
     auto si = py::cast<c10::SymInt>(elm);
-    // TODO: the buffer will need to be made owning later
-    symints.append(si.as_int_unchecked());
+    symints.append(si);
```
Is there any special handling that needs to be done here further down in `values_from_buffer` to handle SymInts? Currently it just does `THPUtils_unpackLongs`.

bdhirsh replied:
whoops, thanks. My PR removes the caching logic anyway, so I'm tentatively just going to delete the `values_from_buffer` code entirely (no reason to stash anything in a python buffer if we aren't caching it).

ezyang replied:
You can't remove values_from_buffer, it is needed to ensure the returned IntArrayRef stays live beyond the extent of the size() call

ezyang commented on tools/autograd/templates/python_variable_methods.cpp:
```diff
@@ -983,8 +983,16 @@ static PyObject * THPVariable_storage(PyObject* self, PyObject* arg)
+  auto curr_type = (PyObject*)Py_TYPE(self);
+  auto curr_type_name = PyObject_GetAttrString(curr_type, "__name__");
+  auto curr_type_str = std::string(PyUnicode_AsUTF8(curr_type_name));
+  auto is_python_functional_tensor = curr_type_str == "FunctionalTensor";
```
nooo please no more lol

ezyang commented on torch/utils/_python_dispatch.py:
```diff
@@ -151,7 +151,99 @@ def transform_subclass(t, callback):
+            # For lists of outputs, need to alias every individual tensor to the input
+            alias_storage(args[0], o)
+
+def return_and_correct_aliasing(func, args, out):
```
docblock here plz

ezyang replied:
It's kind of weird you don't need kwargs 🤔

ezyang commented on torch/_utils_internal.py:
```diff
@@ -64,6 +64,33 @@ def signpost_event(category: str, name: str, parameters: Dict[str, Any]):
+        else:
+            torch._functionalize_sync(t.elem)
+    else:
+        torch._functionalize_sync(t)
```
I'm not entirely sure why this needs to be this complicated

zou3519 commented on torch/utils/_python_dispatch.py:
```diff
@@ -151,7 +151,99 @@ def transform_subclass(t, callback):
+            # For lists of outputs, need to alias every individual tensor to the input
+            alias_storage(args[0], o)
+
+def return_and_correct_aliasing(func, args, out):
```
This is probably worth having tests for

zou3519 commented on torch/utils/_python_dispatch.py:
```diff
@@ -151,7 +151,99 @@ def transform_subclass(t, callback):
+           else:
+               ret.set_(arg.untyped_storage())
+
+    def is_match(arg, ret):
+        arg_aliases = set() if not arg.alias_info else arg.alias_info.before_set
+        out_aliases = set() if not ret.alias_info else ret.alias_info.before_set
+        return len(arg_aliases & out_aliases) > 0
+
+    num_args = len(func._schema.arguments)
+    num_returns = len(func._schema.returns)
+    for arg_idx in range(num_args):
+        for return_idx in range(num_returns):
+            if is_match(func._schema.arguments[arg_idx], func._schema.returns[return_idx]):
+                alias_storage(args[arg_idx], outs[return_idx])
```
NVM, I see it handled below :)

zou3519 commented on torch/utils/_python_dispatch.py:
```diff
@@ -151,7 +151,99 @@ def transform_subclass(t, callback):
+            return before_set[0]
+        return None
+
+    # Fix up the storages of any outs so that they point to the same storage as the input,
+    # if func is a view op.
+    _correct_storage_aliasing(func, args, [out] if not isinstance(out, (list, tuple)) else out)
+
+    # Next: we need to make sure to return inputs directly, if the output is a mutable alias (e.g. add_()).
+
+    # simple case: none of our outputs have mutable aliases, so we can return the output as-is
+    if not any(get_write_alias(r) is not None for r in func._schema.returns):
+        return out
```
Just to make sure I understand what's going on: Let's say we have `sin_(Tensor(a!) x) -> Tensor(a!)`. It looks like we:
1. set the storage on the output to be the same as the input
2. then attempt to return the input
Is that correct? (or does `_correct_storage_aliasing` do something to not set the storage?)

bdhirsh replied:
Coming back to this late, sorry (now that proxy/fake portions of this PR stack is landed, I'm starting to rework the python functionalization PR's).

`_correct_storage_aliasing()` will be a no-op, unless `func` is a view op (its output is a non-mutable alias of an input).

So for an op like `sin_()`, only (2) will happen above. For an op like `aten.view()`, only (1) will happen above.

It's probably technically possible to have to worry about an opt that returns **both** an output that aliases an input, **and** an input itself, but that would probably not work in PT2 for other reasons (functionalization probably wouldn't be able to handle such an op by default)

zou3519 commented on torch/utils/_python_dispatch.py:
```diff
@@ -151,7 +151,99 @@ def transform_subclass(t, callback):
+        arg_idx = get_arg_idx_from_alias(get_write_alias(0))
+        return args[arg_idx]
+
+    return [
+        args[get_arg_idx_from_alias(get_write_alias(0))] for i in range(len(func._schema.returns))
+    ]
```
Do we have an op that does this?

zou3519 commented on torch/utils/_python_dispatch.py:
```diff
@@ -151,7 +151,99 @@ def transform_subclass(t, callback):
+    if not all(get_write_alias(r) is not None for r in func._schema.returns):
+        raise RuntimeError("Unsupported schema: " + str(func._schema))
+
+    if len(func._schema.returns) == 1:
+        arg_idx = get_arg_idx_from_alias(get_write_alias(0))
+        return args[arg_idx]
```
Let's say we have `sin_(Tensor(a!) x) -> Tensor(a!)`. Is the contract that the output must be the exact same as the input, or is it fine if the output is a view of the input (but we have written to both?)

I'm trying to figure out if we return the Tensor as-is is because we understand `sin_` to be an in-place operation, or if this information is encoded in the schema, or both.

zou3519 commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -42,11 +42,45 @@ def __new__(cls, elem):
+            def unwrap(x):
+                return x.elem
+            args_unwrapped = pytree.tree_map_only(FunctionalTensor, unwrap, args)
+            # All metadata accesses should be plumbed to the inner tensor, that way we don't have to worry
+            # about the problem of keeping metadata in sync between the wrapper and inner tensor.
+            # This also alleviates us from having to manually handle metadata mutations on the wrapper.
```
Or another question is: FunctionalTensorWrapper (in C++) doesn't need to do this. Why are we doing it here?

bdhirsh replied:
Sorry this q slipped by.

> gah, is there a way for us to know to update this if we add a new metadata field?

FunctionalTensor should error loudly at least - since `FunctionalTensor.__torch_dispatch__()` always raises an error right now (except for this hardcoded list of metadata requests)

> Or another question is: FunctionalTensorWrapper (in C++) doesn't need to do this. Why are we doing it here?

Actually, FunctionalTensorWrapper does this today ([code](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/FunctionalTensorWrapper.cpp#L370)). It overrides all of the overrideable metadata methods on TensorImpl to just plumb them to the inner tensor.

bdhirsh committed:
Update on "Fix aliasing, metadata and edge case handling for FunctionalTensor, introduce some new wrapper tensor APIs"

This PR provides some updates to the "wrapper tensor" subclass API, that made it possible to integrate with AOTAutograd (next PR in the stack).

Enumeration of changes:

**Changes that are general to wrapper subclasses**

(1) When a subclass specifies `dispatch_sizes_strides_policy="sizes"`, previously it would always cache the result of the first call somewhere on the python tensor. I removed this caching - if we cache, then we won't notice when there is a metadata mutation to the tensor.

(2) Allow subclasses using `dispatch_sizes_strides_policy="sizes"` to work with dynamic shapes. Previously, you'd get an error if you specified a wrapper subclass like the above and tried to give it dynamic sizes. From the comments this sounded like a coverage gap, so I removed the restriction. It also required removing an `as_int_unchecked` call from `ConcretePyInterpreterVTable::sym_sizes`.

(3) Added a new API for wrapper subclasses: `torch.utils._python_dispatch.return_and_correct_aliasing(func, args, out)`.

This API does the work of both (a) aliasing any output storages to inputs when it sees a view op, and (b) making sure that we return inputs directly, when we encounter an inplace op. This should be a one-liner that wrapper subclasses can add to their __torch_dispatch__ code in order to get correct aliasing.

Importantly, a wrapper subclass will **need** to use this API in their subclass to work properly with `torch.compile`.

**Changes specific to FunctionalTensor**

Trying to integrate with AOTAutograd exposed a few bugs in my subclass as written. Stuff that I had to fix included:

(1) Make `torch._sync()` work with `FunctionalTensor`, and make sure to temporarily disable `FunctionalTensorMode` while performing the sync

(2) Use `dispatch_sizes_strides_policy="sizes"`, and have every call to metadata get plumbed to the inner tensor. This is exactly what the C++ `FunctionalTensorWrapper` does today too. Although ezyang points out that at some point we may want to "not trust" the strides of the inner tensor, and compute strides ourselves (FunctionalTensorWrapper... sort of did this a while ago, but we killed it everywhere except for XLA).

(3) Use the new `return_and_correct_aliasing` API.

(4) Make sure to include `DispatchKey.Functionalize` in the TLS, when redispatching in our mode. This ensures that factory ops are hit.




[ghstack-poisoned]

bdhirsh committed:
Update on "Fix aliasing, metadata and edge case handling for FunctionalTensor, introduce some new wrapper tensor APIs"

This PR provides some updates to the "wrapper tensor" subclass API, that made it possible to integrate with AOTAutograd (next PR in the stack).

Enumeration of changes:

**Changes that are general to wrapper subclasses**

(1) When a subclass specifies `dispatch_sizes_strides_policy="sizes"`, previously it would always cache the result of the first call somewhere on the python tensor. I removed this caching - if we cache, then we won't notice when there is a metadata mutation to the tensor.

(2) Allow subclasses using `dispatch_sizes_strides_policy="sizes"` to work with dynamic shapes. Previously, you'd get an error if you specified a wrapper subclass like the above and tried to give it dynamic sizes. From the comments this sounded like a coverage gap, so I removed the restriction. It also required removing an `as_int_unchecked` call from `ConcretePyInterpreterVTable::sym_sizes`.

(3) Added a new API for wrapper subclasses: `torch.utils._python_dispatch.return_and_correct_aliasing(func, args, out)`.

This API does the work of both (a) aliasing any output storages to inputs when it sees a view op, and (b) making sure that we return inputs directly, when we encounter an inplace op. This should be a one-liner that wrapper subclasses can add to their __torch_dispatch__ code in order to get correct aliasing.

Importantly, a wrapper subclass will **need** to use this API in their subclass to work properly with `torch.compile`.

**Changes specific to FunctionalTensor**

Trying to integrate with AOTAutograd exposed a few bugs in my subclass as written. Stuff that I had to fix included:

(1) Make `torch._sync()` work with `FunctionalTensor`, and make sure to temporarily disable `FunctionalTensorMode` while performing the sync

(2) Use `dispatch_sizes_strides_policy="sizes"`, and have every call to metadata get plumbed to the inner tensor. This is exactly what the C++ `FunctionalTensorWrapper` does today too. Although ezyang points out that at some point we may want to "not trust" the strides of the inner tensor, and compute strides ourselves (FunctionalTensorWrapper... sort of did this a while ago, but we killed it everywhere except for XLA).

(3) Use the new `return_and_correct_aliasing` API.

(4) Make sure to include `DispatchKey.Functionalize` in the TLS, when redispatching in our mode. This ensures that factory ops are hit.




[ghstack-poisoned]

bdhirsh committed:
Update on "Fix aliasing, metadata and edge case handling for FunctionalTensor, introduce some new wrapper tensor APIs"

This PR provides some updates to the "wrapper tensor" subclass API, that made it possible to integrate with AOTAutograd (next PR in the stack).

Enumeration of changes:

**Changes that are general to wrapper subclasses**

(1) When a subclass specifies `dispatch_sizes_strides_policy="sizes"`, previously it would always cache the result of the first call somewhere on the python tensor. I removed this caching - if we cache, then we won't notice when there is a metadata mutation to the tensor.

(2) Allow subclasses using `dispatch_sizes_strides_policy="sizes"` to work with dynamic shapes. Previously, you'd get an error if you specified a wrapper subclass like the above and tried to give it dynamic sizes. From the comments this sounded like a coverage gap, so I removed the restriction. It also required removing an `as_int_unchecked` call from `ConcretePyInterpreterVTable::sym_sizes`.

(3) Added a new API for wrapper subclasses: `torch.utils._python_dispatch.return_and_correct_aliasing(func, args, out)`.

This API does the work of both (a) aliasing any output storages to inputs when it sees a view op, and (b) making sure that we return inputs directly, when we encounter an inplace op. This should be a one-liner that wrapper subclasses can add to their __torch_dispatch__ code in order to get correct aliasing.

Importantly, a wrapper subclass will **need** to use this API in their subclass to work properly with `torch.compile`.

**Changes specific to FunctionalTensor**

Trying to integrate with AOTAutograd exposed a few bugs in my subclass as written. Stuff that I had to fix included:

(1) Make `torch._sync()` work with `FunctionalTensor`, and make sure to temporarily disable `FunctionalTensorMode` while performing the sync

(2) Use `dispatch_sizes_strides_policy="sizes"`, and have every call to metadata get plumbed to the inner tensor. This is exactly what the C++ `FunctionalTensorWrapper` does today too. Although ezyang points out that at some point we may want to "not trust" the strides of the inner tensor, and compute strides ourselves (FunctionalTensorWrapper... sort of did this a while ago, but we killed it everywhere except for XLA).

(3) Use the new `return_and_correct_aliasing` API.

(4) Make sure to include `DispatchKey.Functionalize` in the TLS, when redispatching in our mode. This ensures that factory ops are hit.




[ghstack-poisoned]

bdhirsh committed:
Update on "Fix aliasing, metadata and edge case handling for FunctionalTensor, introduce some new wrapper tensor APIs"

This PR provides some updates to the "wrapper tensor" subclass API, that made it possible to integrate with AOTAutograd (next PR in the stack).

Enumeration of changes:

**Changes that are general to wrapper subclasses**

(1) When a subclass specifies `dispatch_sizes_strides_policy="sizes"`, previously it would always cache the result of the first call somewhere on the python tensor. I removed this caching - if we cache, then we won't notice when there is a metadata mutation to the tensor.

(2) Allow subclasses using `dispatch_sizes_strides_policy="sizes"` to work with dynamic shapes. Previously, you'd get an error if you specified a wrapper subclass like the above and tried to give it dynamic sizes. From the comments this sounded like a coverage gap, so I removed the restriction. It also required removing an `as_int_unchecked` call from `ConcretePyInterpreterVTable::sym_sizes`.

(3) Added a new API for wrapper subclasses: `torch.utils._python_dispatch.return_and_correct_aliasing(func, args, out)`.

This API does the work of both (a) aliasing any output storages to inputs when it sees a view op, and (b) making sure that we return inputs directly, when we encounter an inplace op. This should be a one-liner that wrapper subclasses can add to their __torch_dispatch__ code in order to get correct aliasing.

Importantly, a wrapper subclass will **need** to use this API in their subclass to work properly with `torch.compile`.

**Changes specific to FunctionalTensor**

Trying to integrate with AOTAutograd exposed a few bugs in my subclass as written. Stuff that I had to fix included:

(1) Make `torch._sync()` work with `FunctionalTensor`, and make sure to temporarily disable `FunctionalTensorMode` while performing the sync

(2) Use `dispatch_sizes_strides_policy="sizes"`, and have every call to metadata get plumbed to the inner tensor. This is exactly what the C++ `FunctionalTensorWrapper` does today too. Although ezyang points out that at some point we may want to "not trust" the strides of the inner tensor, and compute strides ourselves (FunctionalTensorWrapper... sort of did this a while ago, but we killed it everywhere except for XLA).

(3) Use the new `return_and_correct_aliasing` API.

(4) Make sure to include `DispatchKey.Functionalize` in the TLS, when redispatching in our mode. This ensures that factory ops are hit.




[ghstack-poisoned]

bdhirsh committed:
Update on "Fix aliasing, metadata and edge case handling for FunctionalTensor, introduce some new wrapper tensor APIs"

This PR provides some updates to the "wrapper tensor" subclass API, that made it possible to integrate with AOTAutograd (next PR in the stack).

Enumeration of changes:

**Changes that are general to wrapper subclasses**

(1) When a subclass specifies `dispatch_sizes_strides_policy="sizes"`, previously it would always cache the result of the first call somewhere on the python tensor. I removed this caching - if we cache, then we won't notice when there is a metadata mutation to the tensor.

(2) Allow subclasses using `dispatch_sizes_strides_policy="sizes"` to work with dynamic shapes. Previously, you'd get an error if you specified a wrapper subclass like the above and tried to give it dynamic sizes. From the comments this sounded like a coverage gap, so I removed the restriction. It also required removing an `as_int_unchecked` call from `ConcretePyInterpreterVTable::sym_sizes`.

(3) Added a new API for wrapper subclasses: `torch.utils._python_dispatch.return_and_correct_aliasing(func, args, out)`.

This API does the work of both (a) aliasing any output storages to inputs when it sees a view op, and (b) making sure that we return inputs directly, when we encounter an inplace op. This should be a one-liner that wrapper subclasses can add to their __torch_dispatch__ code in order to get correct aliasing.

Importantly, a wrapper subclass will **need** to use this API in their subclass to work properly with `torch.compile`.

**Changes specific to FunctionalTensor**

Trying to integrate with AOTAutograd exposed a few bugs in my subclass as written. Stuff that I had to fix included:

(1) Make `torch._sync()` work with `FunctionalTensor`, and make sure to temporarily disable `FunctionalTensorMode` while performing the sync

(2) Use `dispatch_sizes_strides_policy="sizes"`, and have every call to metadata get plumbed to the inner tensor. This is exactly what the C++ `FunctionalTensorWrapper` does today too. Although ezyang points out that at some point we may want to "not trust" the strides of the inner tensor, and compute strides ourselves (FunctionalTensorWrapper... sort of did this a while ago, but we killed it everywhere except for XLA).

(3) Use the new `return_and_correct_aliasing` API.

(4) Make sure to include `DispatchKey.Functionalize` in the TLS, when redispatching in our mode. This ensures that factory ops are hit.




[ghstack-poisoned]

bdhirsh committed:
Update on "Fix aliasing, metadata and edge case handling for FunctionalTensor, introduce some new wrapper tensor APIs"

This PR provides some updates to the "wrapper tensor" subclass API, that made it possible to integrate with AOTAutograd (next PR in the stack).

Enumeration of changes:

**Changes that are general to wrapper subclasses**

(1) When a subclass specifies `dispatch_sizes_strides_policy="sizes"`, previously it would always cache the result of the first call somewhere on the python tensor. I removed this caching - if we cache, then we won't notice when there is a metadata mutation to the tensor.

(2) Allow subclasses using `dispatch_sizes_strides_policy="sizes"` to work with dynamic shapes. Previously, you'd get an error if you specified a wrapper subclass like the above and tried to give it dynamic sizes. From the comments this sounded like a coverage gap, so I removed the restriction. It also required removing an `as_int_unchecked` call from `ConcretePyInterpreterVTable::sym_sizes`.

(3) Added a new API for wrapper subclasses: `torch.utils._python_dispatch.return_and_correct_aliasing(func, args, out)`.

This API does the work of both (a) aliasing any output storages to inputs when it sees a view op, and (b) making sure that we return inputs directly, when we encounter an inplace op. This should be a one-liner that wrapper subclasses can add to their __torch_dispatch__ code in order to get correct aliasing.

Importantly, a wrapper subclass will **need** to use this API in their subclass to work properly with `torch.compile`.

**Changes specific to FunctionalTensor**

Trying to integrate with AOTAutograd exposed a few bugs in my subclass as written. Stuff that I had to fix included:

(1) Make `torch._sync()` work with `FunctionalTensor`, and make sure to temporarily disable `FunctionalTensorMode` while performing the sync

(2) Use `dispatch_sizes_strides_policy="sizes"`, and have every call to metadata get plumbed to the inner tensor. This is exactly what the C++ `FunctionalTensorWrapper` does today too. Although ezyang points out that at some point we may want to "not trust" the strides of the inner tensor, and compute strides ourselves (FunctionalTensorWrapper... sort of did this a while ago, but we killed it everywhere except for XLA).

(3) Use the new `return_and_correct_aliasing` API.

(4) Make sure to include `DispatchKey.Functionalize` in the TLS, when redispatching in our mode. This ensures that factory ops are hit.




[ghstack-poisoned]

bdhirsh commented on torch/utils/_python_dispatch.py:
```diff
@@ -151,7 +151,99 @@ def transform_subclass(t, callback):
+
+def _correct_storage_aliasing(func, args, outs):
+    assert isinstance(func, torch._ops.OpOverload)
+    assert isinstance(args, (list, tuple))
```
i think you're right, thanks

bdhirsh commented on torch/utils/_python_dispatch.py:
```diff
@@ -151,7 +151,99 @@ def transform_subclass(t, callback):
+            # in theory if a subclass that needs this API wants to sometimes return
+            # plain tensors, we could remove the assert and just not perform the aliasing,
+            # but it seems safer to learn more about this case first.
+            assert hasattr(x, '__torch_dispatch__')
```
Hmm, starting with that stronger assertion sounds ok to me (if we discover a reason to relax it later we can always wait to understand that situation).

bdhirsh commented on torch/csrc/autograd/python_torch_functions_manual.cpp:
```diff
@@ -566,8 +567,8 @@ static PyMethodDef torch_functions_manual[] = {
-    {"_sync",
-     castPyCFunctionWithKeywords(THPVariable__sync),
+    {"_functionalize_sync",
+     castPyCFunctionWithKeywords(THPVariable__functionalize_sync),
```
will do

soulitzer commented on torch/overrides.py:
```diff
@@ -1905,8 +1905,11 @@ def get_buffer(tensor_subclass, data, prefix):
+    # (2) caching doesn't save us much anyway - we have to go into python every time
+    #     the user calls .shape in any case.
+    SizeType = ctypes.c_longlong * len(data)
+    setattr(tensor_subclass, buffer_name, SizeType(*data))
```
@bdhirsh  I ended up needing to fix this as well to get NestedTensor to properly return a singleton symint through the slow path. 
So in case you haven't gotten the chance to look at this further, I have a PR https://github.com/pytorch/pytorch/pull/107839.

The PR removes `values_from_buffer` as you suggested, but also account for Ed's comment https://github.com/pytorch/pytorch/pull/106405/files#r1282302973

bdhirsh commented:
Closing this PR in favor of https://github.com/pytorch/pytorch/pull/106404 (and the other PR's) around it. I tried to split it up into:

- https://github.com/pytorch/pytorch/pull/106404: adds a `FunctionalTensor` and `FunctionalTensor`Mode
- 2-3 other PR's that fix up various subclass interactions, that aren't really specific to functionalization
